[SPMD] add FQN argument to Override.replacement

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100473

Differential Revision: [D45486089](https://our.internmc.facebook.com/intern/diff/D45486089)

mrshenli committed:
[SPMD] add FQN argument to Override.replacement

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45486089).

wanchaol commented on test/distributed/_spmd/test_tracing.py:
```diff
@@ -516,7 +516,9 @@ class AssertOverride(Override):
 
-            def replacement(self, orig_submodule: torch.nn.Module) -> torch.nn.Module:
+            def replacement(
+                self, orig_submodule: torch.nn.Module, fqn: str
```
nit: should `fqn` be the first argument instead of `orig_submodule`?

mrshenli replied:
good point, let me update.

wanchaol reviewed (approved):
Curious what's the motivation of this patch, is the same instance of `origin_submodule` being shared by multiple fqns, and we only want to switch one fqn pointed submodule?

mrshenli commented:
> Curious what's the motivation of this patch, is the same instance of origin_submodule being shared by multiple fqns, and we only want to switch one fqn pointed submodule?

@yifuwang mentioned a case on the sparse side where there are multiple submodules of the same type sharing the same global sharding plan. We need to know which part of the plan each submodule corresponds to. So having the module instance along is not sufficient, and we use fqn to distinguish them.

mrshenli committed:
Update on "[SPMD] add FQN argument to Override.replacement"


Differential Revision: [D45486089](https://our.internmc.facebook.com/intern/diff/D45486089)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45486089).

mrshenli committed:
Update on "[SPMD] add FQN argument to Override.replacement"


Differential Revision: [D45486089](https://our.internmc.facebook.com/intern/diff/D45486089)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45486089).

mrshenli commented:
@pytorchbot merge

mrshenli commented:
@pytorchbot merge -f "8 CI test envs failed to start"
